src/persist-client/src/operators/shard_source.rs

```diff
@@ -273,21 +273,21 @@ where
 
     // This feels a bit clunky but it makes sure that we can't misuse the error
     // handler below.
-    struct ErrorHandler<H: FnMut(String) -> () + 'static> {
+    struct ErrorHandler<H: FnOnce(String) -> Pin<Box<dyn Future<Output = ()>>> + 'static> {
         inner: H,
     }
-    impl<H: FnMut(String) -> () + 'static> ErrorHandler<H> {
+    impl<H: FnOnce(String) -> Pin<Box<dyn Future<Output = ()>>> + 'static> ErrorHandler<H> {
         /// Report the error and enforce that we never return.
-        async fn report_and_stop<R>(&mut self, error: String) -> R {
-            (self.inner)(error);
+        async fn report_and_stop(self, error: String) {
```

Comment:
Couldn't you make this return `!`, and then get rid of the `unreachable!()`s below?

edit: Ah, that probably doesn't work because it's async?

Reply:
I think it would, but the never type is unstable https://doc.rust-lang.org/std/primitive.never.html